WIP: Kraken buy multiple assets

Sebastix opened the pull request.

jorijn committed:
composer update

jorijn committed:
added kraken settings, removed hardcoded references to EUR to provide for configurable base currencies

jorijn committed:
added services for buying, withdrawing and checking the balance for kraken

Sebastix committed:
development with docker-compose

Sebastix committed:
kraken buy asset by choice and test adjustments

Sebastix committed:
return float balance by asset

jorijn commented on docker/development/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+FROM php:7.4-cli-alpine
```
Tip: Maak hier een multistage Dockerfile van, je kan dan de productie laag van de image hergebruiken voor development.

Voorbeeld:

FROM php:7.4-cli-alpine AS production

productie dingen hier

FROM production AS development

hier development dingen toevoegen

Sebastix replied:
Handy! Ik zag het ook in Dockerfile in de app root voorbij komen, ben niet bekend met het multistage concept binnen Docker maar sounds logic :) Kan dit dan ook in de Dockerfile geplaatst die in de app root staat?

Sebastix replied:
Ik heb nu een development stage toegevoegd in de Dockerfile:

```
##################################################################################################################
# Development Stage
##################################################################################################################
FROM production AS dev

RUN composer install
```

jorijn replied:
Zeker! Je zal dan alleen tijdens het bouwen van bijv. de productiebuild moeten aangeven dat je tot een specifiek target wil gaan en niet verder.

Sebastix replied:
Heb nu een development stage (helemaal onderaan) toegevoegd `FROM base_image AS development` in de Dockerfile in de app root. Hier staan verder (nog) geen commands in.
In `docker/development/docker-compose.yml` heb ik het volgende nu staan mbt de build:
```
    build:
      context: ../../
      dockerfile: Dockerfile
      target: development
```

Als ik echter daarna `docker-compose --build` doe, zou ik verwachten dat er een development image wordt gebouwd voor de service, zonder de test en production_build stages. Toch lijkt dit niet te gebeuren, want ik zie bij het builden de output van test voorbij komen.

Hetzelfde als ik `docker build --target development -t jorijn/bitcoin-dca:development .` uitvoer.

jorijn commented on docker/development/README.md:
````diff
@@ -0,0 +1,52 @@
+### Start development
+First, make sure you start with installing the software dependencies with Composer.  
+```
+docker exec -it dca_php composer install
````
Het afleiden van de container naam kan hier achterwege gelaten worden: `docker-compose exec dev_dca composer install`

Sebastix replied:
Als ik geen `container_name` opgeef in de `docker-compose.yml` dan krijgt de container de naam `development_dca_1` (icm met een development stage in de Dockerfile)

jorijn replied:
Klopt, maar als je `exec` via docker-compose uitvoert kun je de naam van de service gebruiken en is de naam van je container niet meer relevant.

Sebastix replied:
Ah, dan wordt het command `docker exec -it <container_name> <command>` dus `docker-compose exec <service_name> <command>` met als kanttekening dat deze laatste command wel in de `docker/development` directory uitgevoerd moet worden :) Helder en is logischer mbt development!

jorijn commented on docker/development/docker-compose.yml:
```diff
@@ -0,0 +1,11 @@
+      dockerfile: docker/development/Dockerfile
+    container_name: dca_php
+    volumes:
+    - ./.env:/app/.env
```
In de container kun je beter environment vars gebruiken:

```yaml
services:
  dca:
    env_file: ./.env
```

Sebastix replied:
Die property kende ik nog niet thanks :)

jorijn commented on docker/development/docker-compose.yml:
```diff
@@ -0,0 +1,11 @@
+  dev_dca:
+    build:
+      context: ../../
+      dockerfile: docker/development/Dockerfile
```
Is er een speciale reden waarom dit bestand weg verstopt zit in een subdirectory? Zou m.i. prima in project root kunnen.

Sebastix replied:
Nee geen specifieke reden hiervoor, ben het zelf gewend om dit altijd zo te doen in projecten waar ik in werk waar we alles in 1 docker directory werken ipv alles in de root te plaatsen. In deze docker dir plaatsen we dan ook aanvullende files zoals bv config bestanden (php.ini of een settings.php in geval van Drupal).
Het zou hier inderdaad in de app root kunnen staan, al ben ik wel geneigd om alles in een directory Docker te plaatsen voor het overzicht.

jorijn commented on src/Command/BuyCommand.php:
```diff
@@ -56,6 +57,18 @@ public function execute(InputInterface $input, OutputInterface $output): int
 
+        $assetToBuy = (string) $input->getArgument('asset');
+        // TODO check if this a valid token which can be bought on the exchange
+        if ($assetToBuy === '') {
```
Ik weet niet zeker of deze situatie eigenlijk wel kan voorkomen, je hebt "asset" al als REQUIRED aangemerkt, je kan deze dus niet meegeven en toch leeg hebben.

Sebastix replied:
True, die check is niet nodig.

jorijn commented on src/Command/BuyCommand.php:
```diff
@@ -56,6 +57,18 @@ public function execute(InputInterface $input, OutputInterface $output): int
+          return 1;
+        }
+
+        if (ctype_upper($assetToBuy) === false) {
```
Kleine moeite om de gebruiker hierin te ontzorgen;

```php
$assetToBuy = strtoupper($assetToBuy);
```

Scheelt weer een foutmelding :)

jorijn commented on src/Command/BuyCommand.php:
```diff
@@ -56,6 +57,18 @@ public function execute(InputInterface $input, OutputInterface $output): int
         }
 
+        $assetToBuy = (string) $input->getArgument('asset');
+        // TODO check if this a valid token which can be bought on the exchange
```
Goed idee, een publieke methode in de BuyServiceInterface zou een goede plek zijn.

Sebastix replied:
Is toegevoegd in de BuyServiceInterface en de TODO's staan ook in de <Exchange>BuyService files

jorijn commented on src/Service/BuyService.php:
```diff
@@ -35,21 +35,21 @@ public function __construct(
         ];
 
-        $this->logger->info('performing buy for {amount}', $logContext);
+        $this->logger->info('performing buy for {amount} of token {token}', $logContext);
```
Hier wordt {token} gebruikt maar in je context is `asset` toegevoegd.

Sebastix replied:
Inderdaad, een typo

jorijn commented on src/Service/Kraken/KrakenBuyService.php:
```diff
@@ -16,34 +16,54 @@ class KrakenBuyService implements BuyServiceInterface
+   * @return CompletedBuyOrder
+   * @throws PendingBuyOrderException
+   */
+    public function initiateBuy(int $amount, string $asset, bool $simulate = false): CompletedBuyOrder
```
Je wijkt hier af van de interface wat in de regel niet zo'n goed idee is, buiten development is er eigenlijk ook geen reden om de aankoop te simuleren.

Sebastix replied:
De `$simulate` moet er uit inderdaad. Icm Docker die ziet of je in de development stage zit, zouden we dit wel kunnen gebruiken?

jorijn replied:
Als je er op staat om perse een simulate optie er in te houden zou ik een extra service introduceren, deze kan dan de KrakenBuyService extenden of decoraten. 

Door middel van prioriteiten aan je tags mee te geven kun je dan je `KrakenSimulatedBuyService` net boven de andere `KrakenBuyService` zetten. In de `supports` methode kun je dan controleren op de aanwezigheid van bijvoorbeeld `KRAKEN_SIMULATE_BUY` environment variabele. 

Dit zou dan ongeveer zo uitwerken:

`$ KRAKEN_SIMULATE_BUY=1 bin/bitcoin-dca buy 10 --yes`

Sebastix committed:
Refactoring things for multi asset withdrawing from Kraken

Sebastix committed:
Refactoring things for multi asset withdrawing from Kraken

Sebastix committed:
remove this unnecessary check

Sebastix committed:
sanitize assetToBuy string

Sebastix committed:
fix typo

Sebastix committed:
Docker development optimalisaties

Sebastix committed:
add validateAsset method to BuyServiceInterface.php

Sebastix committed:
update .env.dist instruction for withdrawing from Kraken

Sebastix committed:
readme changes for .env file

Sebastix committed:
uppercase asset withdraw

Sebastix committed:
withdraw fees and some optimalizations

Sebastix committed:
get withdraw fee, separate kraken providers and add asset to all providers

Sebastix committed:
fix withdrawfee

Sebastix committed:
use docker-compose exec instead of docker exec

Sebastix committed:
Add Kraken exchange to README

Sebastix committed:
docker target build

Sebastix committed:
docker compose met target werkt niet, terug naar Dockerfile in docker/development

Sebastix committed:
providers with assets + tests

Sebastix commented on tests/Service/WithdrawServiceTest.php:
```diff
@@ -258,7 +274,7 @@ public function testGetRecipientAddress(): void
             $this->logger,
             $this->configuredExchange
-        ))->getRecipientAddress();
+        ))->getRecipientAddress($asset);
```
@Jorijn Het zijn ondertussen aardig wat changes...teveel eigenlijk nu.
Elke WithdrawAddressProvider heb ik voorzien van een van een $asset property, zodat er gecheckt kan worden welk adres hoort bij welke asset (als je een andere suggestie hebt om dit op te lossen hoor ik dat graag).
Echter loop ik vast op 1 ding en dat is hier in de WithdrawServiceTest.php. Ik heb geen ervaring met het schrijven van PHP unit tests, dus ik begrijp niet goed wat ik hier moet oplossen:

```
 ✘ Get recipient address
   ┐
   ├ Expectation failed for method name is "provide" when invoked 2 time(s).
   ├ Method was expected to be called 2 times, actually called 0 times.     
   ┴
```

Sebastix committed:
composer update

Sebastix committed:
withdraw kraken multiple assets rewriting

Sebastix committed:
- back to integers instead of floats
- KrakenWithdrawAddressProvider.php with getAsset function from env key variable

Sebastix committed:
fix simulate

Sebastix committed:
update .env.dist

Sebastix committed:
XBT to XXBT fixes

Sebastix committed:
readme update

Sebastix committed:
remove logo

Sebastix committed:
kraken only

Sebastix committed:
update readme

Sebastix committed:
fix tests + update documentation
